parsing/location.ml

```diff
@@ -292,90 +413,185 @@ let highlight_terminfo lb ppf locs =
-    let line_start, line_end, line_start_pos, line_end_pos =
-      let line_start = ref 0 and line_end = ref 0 in
-      let line_start_pos = ref 0 and line_end_pos = ref 0 in
-      let rec loop pos line line_bol_pos =
-        if pos > end_pos then ()
-        else if read_char pos = '\n' then
-          let line_bol_pos' = pos + 1 in
-          if line_bol_pos <= leftmost_start
-          && leftmost_start < line_bol_pos' then begin
-            line_start := line;
-            line_start_pos := line_bol_pos
-          end;
-          if line_bol_pos <= rightmost_end
-          && rightmost_end < line_bol_pos' then begin
-            line_end := line;
-            line_end_pos := pos
-          end;
-          loop (pos + 1) (line + 1) line_bol_pos'
-        else
-          loop (pos + 1) line line_bol_pos
-      in
-      loop 0 0 0;
-      !line_start, !line_end, !line_start_pos, !line_end_pos
+    let c = Bag.Cursor.create lbag in
+    let lines =
+      get_lines ~start_pos:(Bag.leftmost lbag) ~end_pos:(Bag.rightmost lbag)
+      |> List.map (fun (line, s) ->
+        let _ = Bag.Cursor.advance_at c ~pos:s in
+        let after_line_end = String.length line + 1 in
+        let pos_in_line = Bag.Cursor.advance c ~offset:after_line_end in
+        let line_nb = match pos_in_line with
+          | [] -> ""
+          | (_, p) :: _ -> string_of_int p.pos_lnum in
+        (line, line_nb, s)
+      )
     in
     Format.fprintf ppf "@[<v>";
-    (* Print the input, highlighting the locations.
-       Indent by two spaces. *)
-    Format.fprintf ppf "  @[<v>";
-    if line_start = line_end then begin
-      (* single-line error *)
-      for pos = line_start_pos to line_end_pos - 1 do
-        Format.pp_print_char ppf (read_char pos)
-      done;
-      Format.fprintf ppf "@,";
-      for pos = line_start_pos to rightmost_end - 1 do
-        if is_highlighted pos then Format.pp_print_char ppf '^'
-        else Format.pp_print_char ppf ' '
-      done
-    end else begin
-      (* multi-line error *)
-      for pos = line_start_pos to line_end_pos - 1 do
-        let c = read_char pos in
-        if c = '\n' then Format.fprintf ppf "@,"
-        else if is_highlighted pos then Format.pp_print_char ppf c
-        else Format.pp_print_char ppf '.'
-      done
+    let c = Bag.Cursor.create lbag in
+    begin match lines with
+    | [] | [("", _, _)] -> ()
+    | [(line, line_nb, line_start_cnum)] ->
+        (* Single-line error *)
+        Format.fprintf ppf "%s | %s@," line_nb line;
+        Format.fprintf ppf "%*s   " (String.length line_nb) "";
+        let _ = Bag.Cursor.advance_at c ~pos:line_start_cnum in
+        for pos = line_start_cnum + 1 to (Bag.rightmost lbag).pos_cnum do
+          let ps = Bag.Cursor.advance_at c ~pos in
+          List.iter (function
+            | `Start, _ -> Format.fprintf ppf "@{<%s>" highlight_tag
+            | `End, _ -> Format.fprintf ppf "@}"
+          ) ps;
```

Comment:
Does that really work correctly when there are multiple overlapping locations positions? I might, but it looks a bit subtle.